llvm/lib/Target/RISCV/RISCVInstrInfo.cpp

```diff
@@ -302,102 +302,87 @@ void RISCVInstrInfo::copyPhysRegVector(MachineBasicBlock &MBB,
-  if (forwardCopyWillClobberTuple(DstEncoding, SrcEncoding, NF * LMulVal)) {
-    I = NF - 1;
-    End = -1;
-    Incr = -1;
-  }
-
-  for (; I != End; I += Incr) {
-    auto MIB =
-        BuildMI(MBB, MBBI, DL, get(Opc), TRI->getSubReg(DstReg, SubRegIdx + I));
-    if (UseVMV_V_V)
-      MIB.addReg(TRI->getSubReg(DstReg, SubRegIdx + I), RegState::Undef);
-    if (UseVMV_V_I)
-      MIB = MIB.add(DefMBBI->getOperand(2));
-    else
-      MIB = MIB.addReg(TRI->getSubReg(SrcReg, SubRegIdx + I),
-                       getKillRegState(KillSrc));
-    if (UseVMV_V_V) {
-      const MCInstrDesc &Desc = DefMBBI->getDesc();
-      MIB.add(DefMBBI->getOperand(RISCVII::getVLOpNum(Desc)));  // AVL
-      MIB.add(DefMBBI->getOperand(RISCVII::getSEWOpNum(Desc))); // SEW
-      MIB.addImm(0);                                            // tu, mu
-      MIB.addReg(RISCV::VL, RegState::Implicit);
-      MIB.addReg(RISCV::VTYPE, RegState::Implicit);
+  unsigned NumRegs = NF * LMulVal;
+  bool ReversedCopy =
+      forwardCopyWillClobberTuple(DstEncoding, SrcEncoding, NumRegs);
+
+  unsigned I = 0;
+  while (I != NumRegs) {
+    auto GetCopyInfo =
+        [&](unsigned SrcReg,
+            unsigned DstReg) -> std::tuple<int, const TargetRegisterClass &,
+                                           unsigned, unsigned, unsigned> {
+      unsigned SrcEncoding = TRI->getEncodingValue(SrcReg);
+      unsigned DstEncoding = TRI->getEncodingValue(DstReg);
+      if (!(SrcEncoding & 0b111) && !(DstEncoding & 0b111) && I + 8 <= NumRegs)
+        return {8, RISCV::VRM8RegClass, RISCV::VMV8R_V, RISCV::PseudoVMV_V_V_M8,
+                RISCV::PseudoVMV_V_I_M8};
+      if (!(SrcEncoding & 0b11) && !(DstEncoding & 0b11) && I + 4 <= NumRegs)
+        return {4, RISCV::VRM4RegClass, RISCV::VMV4R_V, RISCV::PseudoVMV_V_V_M4,
+                RISCV::PseudoVMV_V_I_M4};
+      if (!(SrcEncoding & 0b1) && !(DstEncoding & 0b1) && I + 2 <= NumRegs)
+        return {2, RISCV::VRM2RegClass, RISCV::VMV2R_V, RISCV::PseudoVMV_V_V_M2,
+                RISCV::PseudoVMV_V_I_M2};
+      return {1, RISCV::VRRegClass, RISCV::VMV1R_V, RISCV::PseudoVMV_V_V_M1,
+              RISCV::PseudoVMV_V_I_M1};
+    };
+
+    auto [NumCopied, RegClass, Opc, VVOpc, VIOpc] = GetCopyInfo(SrcReg, DstReg);
+
+    MachineBasicBlock::const_iterator DefMBBI;
+    if (isConvertibleToVMV_V_V(STI, MBB, MBBI, DefMBBI, LMul)) {
+      Opc = VVOpc;
+
+      if (DefMBBI->getOpcode() == VIOpc) {
+        Opc = VIOpc;
+      }
     }
+
+    for (MCPhysReg Reg : RegClass.getRegisters()) {
+      if (TRI->getEncodingValue(Reg) == TRI->getEncodingValue(SrcReg)) {
+        SrcReg = Reg;
+        break;
+      }
+    }
+
+    for (MCPhysReg Reg : RegClass.getRegisters()) {
+      if (TRI->getEncodingValue(Reg) == TRI->getEncodingValue(DstReg)) {
+        DstReg = Reg;
+        break;
+      }
+    }
+
+    auto EmitCopy = [&](MCRegister SrcReg, MCRegister DstReg, unsigned Opcode) {
```

Comment:
Why do we need a lambda here. Can't we use this code directly where it's called? There are no returns in it.

Reply:
It was from an old revision when I tried to refactor out this copy function. Removed.